pkg/instancegroups/duplicate.go

```diff
@@ -0,0 +1,146 @@
+
+	duplicate, err := p.GroupUpdate.DuplicateGroup(newName)
+	if err != nil {
+		return update.Errorf("unable to duplicate the instancegroup: %s, error: %v", name, err)
+	}
+
+	// @step: we need to update from cloud knowlegde, not from the spec as runtime might have changed
+	min, max := int32(p.GroupUpdate.CloudGroup.Size()), int32(group.MaxSize)
+	duplicate.Spec.MinSize = &min
+	duplicate.Spec.MaxSize = &max
+
+	// @step: create the new instancegroup and wait for the cluster to validate
+	update.Infof("reconfiguring the cluster with the new instancegroup: %s", newName)
+	if err = update.UpdateCluster(ctx); err != nil {
+		return update.Errorf("unable to update the cluster, error: %v", err)
+	}
+
+	// @step: wait for the instancegroup
+	update.Infof("waiting for instancegroup: %s to scale", newName)
+	if err := p.GroupUpdate.WaitForGroupSize(ctx, newName, int(min), update.ScaleTimeout); err != nil {
+		return update.Errorf("instancegroup: %s has not reached size: %d within: %s", newName, min, update.ScaleTimeout)
+	}
+
+	// @step: get an updated list of instancegroups
+	newList, err := update.Clientset.InstanceGroupsFor(update.Cluster).List(v1.ListOptions{})
+	if err != nil {
+		return update.Errorf("unable to get update list of instancegroups, error: %v", err)
+	}
+
+	// @step: validate the cluster again post creation of new instancegroup
+	update.Infof("validating the cluster post creation of new instancegroup: %s", newName)
+	if err = p.GroupUpdate.ValidateClusterWithTimeout(ctx, newList, update.FailOnValidateTimeout); err != nil {
+		return update.Errorf("unable to validate cluster after %s", update.FailOnValidateTimeout)
+	}
+
+	// @step: if not cloudonly we drain all the nodes and then delete the instance group
+	if !update.CloudOnly && strategy.Drain {
+		options := &DrainOptions{
+			Batch:             strategy.Batch,
+			CloudOnly:         update.CloudOnly,
+			Delete:            false,
+			DrainPods:         strategy.Drain,
+			FailOnValidation:  update.FailOnValidate,
+			Interval:          strategy.Interval.Duration,
+			PostDelay:         strategy.PostDrainDelay.Duration,
+			Timeout:           strategy.DrainTimeout.Duration,
+			ValidateCluster:   !update.CloudOnly,
+			ValidationTimeout: update.FailOnValidateTimeout,
+		}
+		// @step: drain the instancegroup of the pods
+		if err = p.GroupUpdate.DrainGroup(ctx, options, newList); err != nil {
+			return update.Errorf("unable to drain parent instancegroup: %s, error: %v", name, err)
+		}
+		// @step: validate the cluster once more
+		if err = p.GroupUpdate.ValidateClusterWithTimeout(ctx, newList, strategy.Interval.Duration); err != nil {
+			return update.Errorf("unable to validate cluster post drain after %s, error: %v", update.FailOnValidateTimeout, err)
+		}
+	}
+
+	// @step: delete the instancegroup and rename the other one
+	update.Infof("deleting the parent instancegroup: %s from kops cluster", name)
```

Comment:
I'm a little worried that we can't really resume if we're interrupted / crash...